File: Fbs.WebApi/Repository/OtpRepository.cs
using System.Linq.Expressions;
using Fbs.WebApi.Entities;
using Fbs.WebApi.Options;
using Google.Apis.Sheets.v4;
using Google.Apis.Sheets.v4.Data;
using Microsoft.Extensions.Options;

namespace Fbs.WebApi.Repository;

public class OtpRepository(
    InstrumentationSource instrumentation,
    IOptions<GoogleOptions> options,
    SheetsService sheetsService
) : IRepository<Otp>
{
    private readonly string[] _header =
    [
        "Phone", "Code", "CreatedAt"
    ];

    public async Task<List<Otp>> GetListAsync(CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();
        
        var items = await sheetsService.Spreadsheets.Values.Get(options.Value.SpreadsheetId, "OTPs")
            .ExecuteAsync(cancellationToken);

        if (!items.Values.First().SequenceEqual(_header))
        {
            throw new Exception("Unexpected headers for Otp sheet");
        }

        return items.Values
            .Skip(1)
            .Select((row, idx) => new Otp
            {
                Row = idx + 2,
                Phone = row.ElementAtOrDefault(0) as string,
                Code = row.ElementAtOrDefault(1) as string,
                CreatedAt = row.ElementAtOrDefault(2) is string o ? DateTimeOffset.Parse(o) : null,
            })
            .ToList();
    }

    public async Task<Otp?> FindAsync(Expression<Func<Otp, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await GetListAsync(cancellationToken);
        return items.SingleOrDefault(predicate.Compile());
    }

    public async Task<Otp> GetAsync(Expression<Func<Otp, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await GetListAsync(cancellationToken);
        return items.Single(predicate.Compile());
    }

    public async Task<Otp> InsertAsync(Otp entity, CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var r = sheetsService.Spreadsheets.Values.Append(
            new ValueRange
            {
                Values = [[entity.Phone, entity.Code, entity.CreatedAt.ToString()]]
            },
            options.Value.SpreadsheetId,
            "OTPs"
        );

        r.ValueInputOption = SpreadsheetsResource.ValuesResource.AppendRequest.ValueInputOptionEnum.RAW;

        await r.ExecuteAsync(cancellationToken);

        return await GetAsync(o => o.Phone == entity.Phone, cancellationToken);
    }

    public async Task<Otp> UpdateAsync(Otp entity, CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var request = sheetsService.Spreadsheets.Values.Update(
            new ValueRange
            {
                Values = [[entity.Phone, entity.Code, entity.CreatedAt.ToString()]]
            },
            options.Value.SpreadsheetId,
            $"OTPs!A{entity.Row}:C{entity.Row}"
        );

        request.ValueInputOption = SpreadsheetsResource.ValuesResource.UpdateRequest.ValueInputOptionEnum.RAW;
        await request.ExecuteAsync(cancellationToken);

        return entity;
    }

    public async Task DeleteAsync(Expression<Func<Otp, bool>> predicate, CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();
        var entity = await GetAsync(predicate, cancellationToken);

        await sheetsService.Spreadsheets.BatchUpdate(new BatchUpdateSpreadsheetRequest
        {
            Requests =
            [
                new Request
                {
                    DeleteDimension = new DeleteDimensionRequest
                    {
                        Range = new DimensionRange
                        {
                            SheetId = await GetSheetId(cancellationToken),
                            Dimension = "ROWS",
                            StartIndex = entity.Row - 1,
                            EndIndex = entity.Row
                        }
                    }
                }
            ]
        }, options.Value.SpreadsheetId).ExecuteAsync(cancellationToken);
    }

    private async Task<int?> GetSheetId(CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var sheet = await sheetsService.Spreadsheets.Get(options.Value.SpreadsheetId).ExecuteAsync(cancellationToken);
        return sheet.Sheets.Single(s => s.Properties.Title == "OTPs").Properties.SheetId;
    }
}

File: Fbs.WebApi/src/main/kotlin/sg/from/fbs/repository/OtpRepository.kt
package sg.from.fbs.repository

import com.google.api.services.sheets.v4.Sheets
import com.google.api.services.sheets.v4.model.*
import org.springframework.stereotype.Repository
import sg.from.fbs.config.GoogleProperties
import sg.from.fbs.entities.Otp
import java.time.OffsetDateTime

@Repository
class OtpRepository(
    private val sheetsService: Sheets,
    private val options: GoogleProperties
) {

    private val header = listOf("Phone", "Code", "CreatedAt")

    fun getList(): List<Otp> {
        val response = sheetsService.spreadsheets().values()
            .get(options.spreadsheetId, "OTPs")
            .execute()

        val values = response.getValues()
        if (values == null || values.isEmpty()) return emptyList()

        val actualHeader = values.first().map { it.toString() }
        if (actualHeader != header) {
            throw Exception("Unexpected headers for Otp sheet")
        }

        return values.drop(1).mapIndexed { idx, row ->
            Otp(
                row = idx + 2,
                phone = row.getOrNull(0) as? String,
                code = row.getOrNull(1) as? String,
                createdAt = (row.getOrNull(2) as? String)?.let { OffsetDateTime.parse(it) }
            )
        }
    }

    fun find(predicate: (Otp) -> Boolean): Otp? {
        return getList().find(predicate)
    }

    fun get(predicate: (Otp) -> Boolean): Otp {
        return getList().first(predicate)
    }

    fun insert(entity: Otp): Otp {
        val body = ValueRange().setValues(
            listOf(
                listOf(
                    entity.phone,
                    entity.code,
                    entity.createdAt.toString()
                )
            )
        )

        sheetsService.spreadsheets().values()
            .append(options.spreadsheetId, "OTPs", body)
            .setValueInputOption("RAW")
            .execute()

        return get { it.phone == entity.phone }
    }

    fun update(entity: Otp): Otp {
        val body = ValueRange().setValues(
            listOf(
                listOf(
                    entity.phone,
                    entity.code,
                    entity.createdAt.toString()
                )
            )
        )

        sheetsService.spreadsheets().values()
            .update(options.spreadsheetId, "OTPs!A${entity.row}:C${entity.row}", body)
            .setValueInputOption("RAW")
            .execute()

        return entity
    }

    fun delete(predicate: (Otp) -> Boolean) {
        val entity = get(predicate)
        val sheetId = getSheetId()

        val request = Request().setDeleteDimension(
            DeleteDimensionRequest().setRange(
                DimensionRange()
                    .setSheetId(sheetId)
                    .setDimension("ROWS")
                    .setStartIndex(entity.row - 1)
                    .setEndIndex(entity.row)
            )
        )

        sheetsService.spreadsheets().batchUpdate(
            options.spreadsheetId,
            BatchUpdateSpreadsheetRequest().setRequests(listOf(request))
        ).execute()
    }

    private fun getSheetId(): Int {
        val spreadsheet = sheetsService.spreadsheets().get(options.spreadsheetId).execute()
        val sheet = spreadsheet.sheets.single { it.properties.title == "OTPs" }
        return sheet.properties.sheetId
    }
}